// sym-util/index.js
var SymUtil = (function(){
	
	/* Authentication Manager for SymHub */
	var AuthManager = (function(){
		var username = "senthilkumar_muthuve";
		var password = new Buffer("ZHAyMUJ1ZEBzaw==", 'base64').toString('binary');
		var authData = username + ":"+ password;
		var AuthToken = new Buffer(authData).toString('base64');
		var getToken = function(){
			return AuthToken;
		}
		return {
			username : username,
			getToken : getToken
		}
	})();


	

	/* function to handle URL Requests */
	var URLManager = (function(){
		var hostName, pathName, queryString, projectID, sprintID, callMethod;

		var getConfluenceURL = function(start,end){
			var start = (new Date(start)).toISOString().replace(".000","");
			start = start.replace(/(:)/g,"%3A");
			var end = (new Date(end)).toISOString().replace(".000","");
			end = end.replace(/(:)/g,"%3A");
			var params = {
				"subCalendarId" : "120c48c7-036a-465b-b04c-690d5f185361",
				"userTimeZoneId" : "America%2FLos_Angeles",
				"start" : start,
				"end" : end
			}

			hostName = "confluence.ges.symantec.com";
			pathName = "/rest/calendar-services/1.0/calendar/events.json";

			return CommonUtils.generateURL(hostName,pathName,params);
		}

		var getJIRAURL = function(type){

			hostName = "jira.ges.symantec.com";

			/* gets url for projects */
			var getProjectURL = function(){
				pathName = "/rest/greenhopper/latest/sprintquery/";
				pathName = pathName.concat(projectID);
				var params = {
					"includeHistoricSprints" : "true",
					"includeFutureSprints" : "true"
				};
				return CommonUtils.generateURL(hostName,pathName,params);
			}

			/* gets URL for sprints */
			var getSprintURL = function(){
				pathName = "/rest/greenhopper/latest/rapid/charts/sprintreport";
				var params = {
					"rapidViewId" : projectID,
					"sprintId" : sprintID
				}
				return CommonUtils.generateURL(hostName,pathName,params);
			}

			if(type == "project")
				return getProjectURL();
			if(type == "sprint"){
				return getSprintURL();
			}

			return {
				getProjectURL : getProjectURL,
				getSprintURL : getSprintURL
			}

		}


		

		return {
			projectID : projectID,
			sprintID : sprintID,
			getJIRAURL : getJIRAURL,
			getConfluenceURL : getConfluenceURL
		}
	})();


	var CommonUtils = (function(){
		/* converts Object URL to string URL
			@ input : 	{
							hostName : "jira.ges.symantec.com",
							pathName : "/rest/greenhopper/latest/sprintquery/2023",
							params : {
									"includeHistoricSprints" : true,
									"includeFutureSprints" : true
						}
			@output : https://jira.ges.symantec.com/rest/greenhopper/latest/sprintquery/2023?includeHistoricSprints=true&includeFutureSprints=true
		 */
		var convertObjtoURL = function(obj){
			var strURL = "";
			strURL += "https://"
			if(obj.hasOwnProperty("hostName")){
				strURL += obj.hostName;
			}
			if(obj.hasOwnProperty("pathName")){
				strURL += obj.pathName;
			}
			if(obj.hasOwnProperty("params")){
				var queryString = objToQueryString(obj.params);
				strURL += queryString;
			}
			return strURL;
		}

		/* converts obj to URL */
		var generateURL = function(hostName,pathName,params){
			var objURL = {
				hostName : hostName,
				pathName : pathName, 
				params : params
			}
			return convertObjtoURL(objURL);
		}

		/* converts param object to queryString */
		var objToQueryString = function(obj){
			var str = "?";
			for(key of Object.keys(obj)){
				str += key+ "="+ obj[key]+ "&";
			}
			return str.substring(0,str.length - 1);
		}

		return {
			convertObjtoURL : convertObjtoURL,
			generateURL : generateURL,
			objToQueryString : objToQueryString
		}
	})();

	return {
		AuthManager : AuthManager,
		URLManager : URLManager
	}
})();

module.exports = SymUtil;
